### PR TITLE
Object Rearrangement: Pick/place changes for using mesh receptacles

### PR DESCRIPTION
**Pull request by yvsriram**
## Motivation and Context

Earlier we were using receptacles with AABB bounding boxes for the task. We will be switching to receptacles that are represented as triangular meshes. This PR contains updates to policy training for using these new receptacles.

## How Has This Been Tested

Testing on episodes generated using FloorPlanner scenes and GSO/ABO objects.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by yvsriram** [comment]
The review for this can be started once PRs and are merged to object_rearrange branch

**Comment by yvsriram** [comment]
The changes here will make sense once PRs [1047](https://github.com/facebookresearch/habitat-lab/pull/1047) and [1050](https://github.com/facebookresearch/habitat-lab/pull/1050) are merged into `object_rearrange`.

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py [hunk 1]
```diff
@@ -106,6 +107,16 @@ def get_local_surface_center(
...
 
+    def check_if_point_on_surface(
+        self,
+        sim: habitat_sim.Simulator,
+        point: mn.Vector3,
+        threshold: float = 0.05,
+    ) -> bool:
+        """
+        Check if point lies within a `threshold` distance of the receptacle's surface
+        """
```
Comment (ykarmesh) [inline habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:118]: don't we need a `pass` command here?

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py [hunk 2]
```diff
@@ -254,6 +264,20 @@ def get_local_surface_center(
...
+        sim: habitat_sim.Simulator,
+        point: mn.Vector3,
+        threshold: float = 0.05,
+    ) -> bool:
+        local_point = self.get_local_transform(sim).transform_point(point)
+        bounds = self.bounds
+        return (
+            bounds.x().contains(local_point.x)
+            and bounds.z().contains(local_point.z)
+            and np.abs(bounds.y().min - local_point.y) < threshold
```
Comment (ykarmesh) [inline habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:278]: Do we consider y as the up direction?

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py [hunk 3]
```diff
@@ -91,6 +89,9 @@ def get_global_transform(self, sim: habitat_sim.Simulator) -> mn.Matrix4:
                 self.parent_link
             ).absolute_transformation()
 
+    def get_local_transform(self, sim: habitat_sim.Simulator) -> mn.Matrix4:
```
Comment (cpaxton) [inline habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:92]: add a description of this function

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py [hunk 4]
```diff
@@ -254,6 +264,20 @@ def get_local_surface_center(
         local_center.y = self.bounds.y().min
         return local_center
 
+    def check_if_point_on_surface(
```
Comment (cpaxton) [inline habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:267]: add a description please

Comment (cpaxton) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:267]: although this one is pretty straightforward

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py [hunk 5]
```diff
@@ -382,6 +409,22 @@ def find_ge(a, x):
         tri_index = find_ge(self.area_weighted_accumulator, sample_val)
         return tri_index
 
+    def get_local_surface_center(
+        self, sim: habitat_sim.Simulator
+    ) -> mn.Vector3:
+        return self.trimesh.centroid
+
+    def check_if_point_on_surface(
+        self,
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:418]: are you assuming each receptacle has 1 surface?

**Review by cpaxton — COMMENTED** [review]
Minor comments for clarity, otherwise seems good to me

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py (hunk 1 above)
Comment (yvsriram) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:118]: Yeah, it is not needed if we have a docstring - [check this](https://stackoverflow.com/questions/46146044/python-s-empty-function-does-not-require-a-pass-statement). I saw it being done this way in other places in this file.

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py (hunk 2 above)
Comment (yvsriram) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:278]: Thanks for catching this. Yes, y is the up direction but I need to use `self.up_axis`. Will fix this.

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yvsriram) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:418]: Yes, a "Receptacle" here is supposed to correspond to a single surface which is specified using a triangular mesh. A shelf, for example, would have multiple instances of this "Receptacle" for each of its surfaces.

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py (hunk 1 above)
Comment (cpaxton) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:118]: not with the comment, but good for code style i think

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py (hunk 3 above)
Comment (yvsriram) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:92]: Added now

File: habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py (hunk 2 above)
Comment (yvsriram) [reply habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py:278]: Updated

**Review by ykarmesh — APPROVED** [review]

**Pull request merged**